drgn/cli.py

```diff
@@ -198,75 +209,125 @@ def main() -> None:
-            "reinterpret",
-            "sizeof",
-        ]
-        for attr in drgn_globals:
-            init_globals[attr] = getattr(drgn, attr)
-        init_globals["__name__"] = "__main__"
-        init_globals["__doc__"] = None
-
-        histfile = os.path.expanduser("~/.drgn_history")
-        try:
-            readline.read_history_file(histfile)
-        except OSError as e:
-            if not isinstance(e, FileNotFoundError) and not args.quiet:
-                print("could not read history:", str(e), file=sys.stderr)
 
-        def write_history_file() -> None:
-            try:
-                readline.write_history_file(histfile)
-            except OSError as e:
-                if not args.quiet:
-                    print("could not write history:", str(e), file=sys.stderr)
+        def banner_func(banner: str) -> str:
+            if missing_debug_info_warning is not None:
+                return f"{banner}\n{missing_debug_info_warning}"
+            else:
+                return banner
+
+        run_interactive(prog, banner_func=banner_func, quiet=args.quiet)
+
+
+def run_interactive(
+    prog: drgn.Program,
+    banner_func: Optional[Callable[[str], str]] = None,
+    globals_func: Optional[Callable[[Dict[str, Any]], Dict[str, Any]]] = None,
+    quiet: bool = False,
+) -> None:
+    """
+    Run the drgn interactive command line interface.
+
+    This function allows your application to embed the same REPL drgn provides
+    when it is run on the command line in interactive mode. Simply provide an
+    already configured :class:`drgn.Program`, and optionally customize the
+    banner and/or globals, and this function will create the REPL and run it
+    until the user exits.
+
+    If you are using this function to embed within another application using
+    readline, then you'll need to take special care. While drgn tries to clean
+    up any modifications to global state made during this function, it cannot
+    save and restore the readline settings. In particular, drgn clears and uses
+    the readline history, and it binds a custom completer to the TAB key. You
+    should be sure to save your application's history and clear any custom
+    keybindings or completers prior to calling this function, and then re-read
+    your application's history and restore keybindings and completion functions
+    on return.
+
+    :param prog: The program to insert into the CLI namespace as ``prog``
+    :param banner_func: If provided, this function is called with the banner
+      that would normally be printed, and is free to return a modified or new
+      banner instead.
+    :param globals_func: If provided, this function is called with the globals
+      dictionary that would normally be provided to the interactive session.
+      The function may return a new dictionary with any modifications.
+    :param quiet: When true, suppress any warnings related to non-fatal errors.
+    """
+    version = _version_header()
```

Comment:
Nitpick: this is only used in one place in this function, so just put the call below where we use it.
```suggestion
```